### PR TITLE
Fix the display format for file sizes

### PR DESCRIPTION
**Pull request by miroim**
## Description

Replace `humansize` with a function `format_file_size` in `src/directory_listing.rs`
## Related Issue

https://github.com/static-web-server/static-web-server/issues/370

## How Has This Been Tested?

Add some test cases in `src/directory_listing.rs`
Compared and tested with local files

### CODE REVIEW

File: src/directory_listing.rs [hunk 1]
```diff
@@ -589,3 +588,70 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
...
+
+/// Formats the file size in bytes to a human-readable string
+fn format_file_size(size: u64) -> String {
+    let units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"];
+    let mut size_tmp = size;
+
+    for i in 0..units.len() - 1 {
+        if size_tmp < 1024 {
+            // return the size with the current unit
+            return format!("{} {}", size_tmp, units[i]);
```
Comment (palant) [inline src/directory_listing.rs]: This case doesn’t need to be inside the loop – you can have bytes as a special case before it.

File: src/directory_listing.rs [hunk 2]
```diff
@@ -589,3 +588,70 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
         )),
     }
 }
+
+/// Formats the file size in bytes to a human-readable string
+fn format_file_size(size: u64) -> String {
+    let units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"];
```
Comment (palant) [inline src/directory_listing.rs]: This should be a constant.

File: src/directory_listing.rs [hunk 3]
```diff
@@ -589,3 +588,70 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
...
+            return format!("{} {}", size_tmp, units[i]);
+        } else if size_tmp < 1024 * 1024 && i < units.len() - 1 {
+            // return the size divided by 1024 with the next unit
+            return format!("{:.2} {}", size_tmp as f64 / 1024.0, units[i + 1]);
+        } 
+        size_tmp >>= 10;
+    }
+
+    // if size is too large, return the largest unit
+    format!("{} {}", size_tmp, units[units.len() - 1])
```
Comment (palant) [inline src/directory_listing.rs]: We are going to have at most 4096 PiB in 64 bit, not large enough to strike the decimal points. So it’s probably best if we only iterate over `units[1 . .units.len() - 1]` – `units[0]` is handled as a special case up front and `units[units.len() - 1]` is a fallback unit to be used if the loop finishes without determining a suitable unit. You can really iterate over the values, not the indexes in that setup. And you’ll leave the loop with a value that you can divide by 1024.0 to display it.

File: src/directory_listing.rs [hunk 4]
```diff
@@ -589,3 +588,70 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
...
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::directory_listing::format_file_size;
```
Comment (palant) [inline src/directory_listing.rs]: Nit: `use super::format_file_size;`

File: src/directory_listing.rs [hunk 5]
```diff
@@ -589,3 +588,70 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
...
+
+    #[test]
+    fn handle_byte() {
+        let size = 128;
+        assert_eq!("128 B", format_file_size(size))
+    }
+    
+    #[test]
+    fn handle_kibibyte() {
+        let size = 1690;
```
Comment (palant) [inline src/directory_listing.rs]: Nit: Maybe test 1024 just to see that trailing zeros aren’t removed?

**Review by palant — CHANGES_REQUESTED** [review]
Looking good, and I do feel sorry about all the nitpicking.

**Comment by miroim** [comment]
> Looking good, and I do feel sorry about all the nitpicking.

It's all part of helping me improve, and I appreciate your attention to detail.

File: src/directory_listing.rs [hunk 6]
```diff
@@ -555,3 +554,74 @@ fn parse_last_modified(modified: SystemTime) -> Result<DateTime<Local>> {
...
 }
+
+/// Formats the file size in bytes to a human-readable string
+fn format_file_size(size: u64) -> String {
+    const UNITS: [&str; 6] = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"];
+    let mut size_tmp = size;
+
+    if size_tmp < 1024 {
+        // return the size with Byte
+        return format!("{} {}", size_tmp, "B");
```
Comment (palant) [inline src/directory_listing.rs]: Please use `UNITS[0]` as parameter here.

**Review by palant — CHANGES_REQUESTED** [review]
Looks good, one tiny thing left to improve.

**Review by palant — APPROVED** [review]
LGTM

**Review by joseluisq — APPROVED** [review]
Thanks!

**Comment by palant** [comment]
@miroim FYI, there was an issue with the code that clippy complained about (you have to run it with `--tests` command line flag). I fix it in #382.

**Pull request merged**